File: problemA.py
from math import ceil

r, c = input().split(' ')
r, c = int(r), int(c)


def indexToCoord(x):
    return ceil(x / c), (x - 1) % c + 1


def newPosition(_row, _p):
    return indexToCoord(r * (_p - 1) + _row)


def distance(_i, _j):
    return abs(_i - (r+1) / 2) + abs(_j - (c+1) / 2)


p, i, j, s = 0, 0, 0, 0


def findStablePosition(_p):
    stablePos = None
    maxCount = 0

    for row in range(1, r + 1):
        # for every row calculate stable position and number of iterations required

        pos = newPosition(row, _p)
        count = 1

        while True:
            newPos = newPosition(pos[0], _p)

            if newPos == pos:  # if equilibrium is reached
                # if stablePos has been found before and it is not the same as this
                if stablePos and stablePos != newPos:
                    return 0, 0, 0
                else:
                    stablePos = newPos  # if stablePos not found yet

                if count > maxCount:
                    maxCount = count

                break

            pos = (newPos[0], newPos[1])
            count += 1

    return *stablePos, maxCount


for p_ in range(1, c + 1):
    sol = findStablePosition(p_)

    i_, j_, s_ = findStablePosition(p_)
    if distance(i_, j_) < distance(i, j):
        p, i, j, s = p_, i_, j_, s_


print(p, i, j, s)

File: problemA.cxx
// Compile: g++ -g -O2 -std=gnu++17 -static problemA.cxx
// Run: ./a.out
#include <iostream>
#include <math.h>

int r, c;

struct Coord {
    int r;
    int c;
};


Coord indexToCoord(int x) {
    return {ceil(float(x) / c), (x - 1) % c + 1};
}

Coord newPosition(int row, int p) {
    return indexToCoord(r * (p - 1) + row);
}

int distance(Coord pos) {
    return abs(pos.r - (r + 1) / 2.0) + abs(pos.c - (c + 1) / 2.0);
}

void findStablePosOfRow(int row, int p, Coord& posReturn, int& countReturn) {
    Coord pos = newPosition(row, p);
    int count = 1;

    while (true) {
        Coord newPos = newPosition(pos.r, p);

        if ((newPos.r == pos.r) && (newPos.c == pos.c)) {
            posReturn.r = pos.r;
            posReturn.c = pos.c;
            countReturn = count;
            break;
        }

        pos.r = newPos.r;
        pos.c = newPos.c;
        count += 1;
    }
}


void findStablePos(int p, Coord& posReturn, int& countReturn) {
    Coord stablePos;
    int maxCount;

    findStablePosOfRow(1, p, stablePos, maxCount);

    for (int row = 2; row <= r; row++) {
        Coord pos;
        int count;
        findStablePosOfRow(row, p, pos, count);

        if (pos.r != stablePos.r || pos.c != stablePos.c) {
            posReturn.r = 0;
            posReturn.c = 0;
            countReturn = 0;
            return;
        }

        if (count > maxCount) {
            maxCount = count;
        }
    }

    posReturn.r = stablePos.r;
    posReturn.c = stablePos.c;
    countReturn = maxCount;
}


int main() {
    std::cin >> r >> c;

    int p=0, i=0, j=0, s=0;
    
    for (int p_ = 1; p_ <= c; p_++) {
        Coord pos;
        int count;
        findStablePos(p_, pos, count);
        if (distance(pos) < distance({i, j})) {
            p = p_;
            i = pos.r;
            j = pos.c;
            s = count;
        }
    }

    std::cout << p << " " << i << " " << j << " " << s << std::endl;
}
